File: backend/src/main/java/com/ssafy/ssafit/service/ReviewService.java
package com.ssafy.ssafit.service;

import com.ssafy.ssafit.dao.ReviewMapper;
import com.ssafy.ssafit.domain.Review;
import com.ssafy.ssafit.dto.request.ReviewRequestDTO;
import com.ssafy.ssafit.dto.response.ReviewResponseVO;
import com.ssafy.ssafit.exception.MemberNotAuthenticatedException;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
public class ReviewService {

    private final ReviewMapper reviewMapper;

    public ReviewService(ReviewMapper reviewMapper) {
        this.reviewMapper = reviewMapper;
    }

    // 리뷰 생성
    @Transactional
    public ReviewResponseVO createReview(ReviewRequestDTO reviewRequestDTO) {
        // 사용자 ID를 인증 정보에서 가져온다고 가정
        String memberId = getAuthenticatedMemberId();

        Review review = Review.builder()
                .reviewId(UUID.randomUUID().toString())
                .memberId(memberId)
                .videoId(reviewRequestDTO.getVideoId())
                .content(reviewRequestDTO.getContent())
                .rating(reviewRequestDTO.getRating())
                .createdAt(new java.util.Date())
                .build();

        reviewMapper.insertReview(review);

        ReviewResponseVO reviewResponseVO = ReviewResponseVO.builder().reviewId(review.getReviewId())
                                            .memberId(review.getMemberId())
                                            .videoId(review.getVideoId()).content(review.getContent())
                                            .rating(review.getRating()).createdAt(review.getCreatedAt())
                                            .build();
        return reviewResponseVO;
    }

    // 특정 영상에 대한 리뷰 리스트 조회
    public List<ReviewResponseVO> getReviewsByVideoId(String videoId) {
        List<Review> reviews = reviewMapper.selectReviewsByVideoId(videoId);
        return reviews.stream()
                .map(ReviewResponseVO::new)
                .toList();
    }

    // 리뷰 수정
    @Transactional
    public ReviewResponseVO updateReview(String reviewId, ReviewRequestDTO reviewRequestDTO) {
        Review review = reviewMapper.selectReviewById(reviewId);
        if (review == null) {
            throw new RuntimeException("Review not found");
        }

        review.setContent(reviewRequestDTO.getContent());
        review.setVideoId(reviewRequestDTO.getVideoId());
        review.setRating(reviewRequestDTO.getRating());

        reviewMapper.updateReview(review);
        return new ReviewResponseVO(review);
    }

    // 리뷰 삭제
    @Transactional
    public void deleteReview(String reviewId) {
        reviewMapper.deleteReview(reviewId);
    }

    // 인증된 사용자 ID 가져오기
    public String getAuthenticatedMemberId() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication == null || !(authentication.getPrincipal() instanceof String)) {
            throw new MemberNotAuthenticatedException("Member not authenticated");
        }
        return (String) authentication.getPrincipal();
    }
}

File: frontend/src/api/services/ReviewService.ts
import api from '../apiClient'
import type { ReviewRequestDTO } from '../interfaces/request'
import type { ReviewResponseVO } from '../interfaces/response'

class ReviewService {
  private static path = '/api/reviews'

  // 리뷰 생성
  static async createReview(reviewRequestDTO: ReviewRequestDTO): Promise<ReviewResponseVO> {
    const response = await api.post(this.path, reviewRequestDTO)
    return response.data
  }

  static async getReviews(): Promise<ReviewResponseVO> {
    const response = await api.get(this.path + '/member')
    return response.data
  }

  static async getReviewsByMemberId(memberId: string): Promise<ReviewResponseVO> {
    const response = await api.get(this.path + `/member/${memberId}`)
    console.log('남의 리뷰' + response.data)
    return response.data
  }

  static async getReviewsByMyId(): Promise<ReviewResponseVO> {
    const response = await api.get(this.path + '/video')
    return response.data
  }
 

  static async getReviewsByVideoId(videoId: string): Promise<ReviewResponseVO[]> {
    const response = await api.get(`${this.path}/list/${videoId}`)
    return response.data
  }

  // 리뷰 수정
  static async updateReview(
    reviewId: string,
    reviewRequestDTO: ReviewRequestDTO,
  ): Promise<ReviewResponseVO> {
    const response = await api.put(`${this.path}/${reviewId}`, reviewRequestDTO)
    return response.data
  }

  // 리뷰 삭제
  static async deleteReview(reviewId: string): Promise<void> {
    await api.delete(`${this.path}/${reviewId}`)
  }
}

export default ReviewService
